core/src/test/scala/cats/derived/contravariant.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2015 Miles Sabin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats
+package derived
+
+import cats.instances.all._
+import cats.laws.discipline.{ContravariantTests, SerializableTests}
+import org.scalacheck.Gen
+import org.scalacheck.Gen.Choose
+
+class ContravariantSuite extends KittensSuite {
+  import TestDefns._
+  import TestEqInstances._
+
+  type OptPred[A] = Option[A => Boolean]
+  type ListPred[A] = List[A => Boolean]
+  type GenericAdtF[A] = GenericAdt[A => Boolean]
+  type ListFToInt[A] = List[Snoc[A => Int]]
+  type InterleavedF[A] = Interleaved[A => Int]
+  type AndCharF[A] = (A => Int, Char)
+  type TreeF[A] = Tree[A => Int]
+
+  case class Pred[A](run: A => Boolean)
+  case class TwiceNest[A](run: Pred[Pred[A]])
+  case class ThriceNest[A](run: Pred[TwiceNest[A]])
+
+  implicit def eqIntPred[A: Numeric: Choose, B: Eq]: Eq[A => B] = new Eq[A => B] {
+    def diagonal[A](a: A): (A, A) = (a, a)
+    override def eqv(x: A => B, y: A => B): Boolean =
+      Gen.posNum[A]
+        .sample
+        .exists(diagonal[A] _ andThen (x *** y) andThen (Eq[B].eqv _).tupled)
```

Comment:
It's better to use `MiniInt` and `Boolean` instead of `Int` and `Long` for `ContravariantTests` instead of defining this awkward `Eq` instances for functions. `cats-laws` had equality for functions based on `Arbitrary` but it was removed and I think we shouldn't introduce it if we can avoid it.

Reply:
Gotcha, nice, yeah I imagined there was something better than this.